### PR TITLE
upgrade spmd_types to 0.2.5, remove SpmdLayout

### PR DESCRIPTION
**Pull request by pianpwk**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* #4325
* #4311
* __->__ #4265

### CODE REVIEW

File: torchtitan/components/loss.py [hunk 1]
```diff
@@ -261,13 +252,14 @@ def __call__(
         """Return the scaled loss and any metrics computed by the loss."""
         del kwargs
         loss = self.fn(pred, labels)
+        if get_spmd_backend() == "spmd_types":  # loss: V->P, annotate global_valid_tokens
+            spmd.assert_type(loss, {"dp": spmd.P, "cp": spmd.P})
+            spmd.assert_type(
+                global_valid_tokens,
+                {"dp": spmd.R, "cp": spmd.R, "tp": spmd.I},
+            )
```
Comment (tianyu-l) [inline torchtitan/components/loss.py]: sometimes you only do assert type when type checking is on -- what's the principle?

File: torchtitan/distributed/parallel_dims.py [hunk 2]
```diff
@@ -52,67 +53,15 @@ class MeshAxisName(StrEnum):
     EFSDP = "efsdp"
 
 
-@dataclass(frozen=True, slots=True)
-class SpmdLayout:
-    """Temporary SPMD layout annotations keyed by logical mesh axis name.
+SpmdLayout = spmd.SpmdType
```
Comment (tianyu-l) [inline torchtitan/distributed/parallel_dims.py]: can we deprecate this indirection of typing and directly use SpmdType?

File: torchtitan/distributed/parallel_dims.py [hunk 3]
```diff
@@ -52,67 +53,15 @@ class MeshAxisName(StrEnum):
...
-                axes = entry if isinstance(entry, tuple) else (entry,)
-                for axis_name in axes:
-                    if not isinstance(axis_name, MeshAxisName):
-                        raise TypeError(
-                            f"Expected MeshAxisName in partition_spec, "
-                            f"got {axis_name!r}."
-                        )
-                    result[axis_name] = spmd.S(dim)
-        return result
+def layout_axes(layout: SpmdLayout) -> tuple[MeshAxisName, ...]:
```
Comment (tianyu-l) [inline torchtitan/distributed/parallel_dims.py]: feels like should be in spmd_types repo?

File: torchtitan/components/loss.py (hunk 1 above)
Comment (pianpwk) [reply torchtitan/components/loss.py]: ah, one thing is that we have single-CPU unit tests on some modules, and assert_type will look for the ambient mesh to reference named axes against. We don't have mesh setup for these tests, so "dp/cp/tp" don't exist for single CPU.

We could set it up (fake backend), but it feels a bit overkill and more machinery for generic module use

File: torchtitan/distributed/parallel_dims.py (hunk 3 above)
Comment (pianpwk) [reply torchtitan/distributed/parallel_dims.py]: yes, probably just missed this one as a helper

File: torchtitan/distributed/parallel_dims.py (hunk 2 above)
Comment (pianpwk) [reply torchtitan/distributed/parallel_dims.py]: sure, just trying to reduce blast radius of PR

File: torchtitan/distributed/parallel_dims.py [hunk 4]
```diff
@@ -51,67 +51,16 @@ class MeshAxisName(StrEnum):
...
-                axes = entry if isinstance(entry, tuple) else (entry,)
-                for axis_name in axes:
-                    if not isinstance(axis_name, MeshAxisName):
-                        raise TypeError(
-                            f"Expected MeshAxisName in partition_spec, "
-                            f"got {axis_name!r}."
-                        )
-                    result[axis_name] = spmd.S(dim)
-        return result
+def layout_axes(layout: SpmdType) -> tuple[MeshAxisName, ...]:
```
Comment (tianyu-l) [inline torchtitan/distributed/parallel_dims.py]: 1. should be put in spmd_types.py
2. should rename because "layout" is gone; maybe to `spmd_types_axes` or `spmd_axes`?

File: torchtitan/distributed/spmd_types.py [hunk 5]
```diff
@@ -190,23 +186,23 @@ def annotate_input_spmd_types(
     ``S(0)@DP, S(0)@CP, R@TP``; labels are
     ``S(0)@DP, S(0)@CP, I@TP``.
     """
-    from torchtitan.distributed.parallel_dims import MeshAxisName
+    from torchtitan.distributed.parallel_dims import MeshAxisName as _MeshAxisName
```
Comment (tianyu-l) [inline torchtitan/distributed/spmd_types.py]: oh what's going on

File: torchtitan/models/common/attention.py [hunk 6]
```diff
@@ -87,6 +86,21 @@ class VarlenMetadata(NamedTuple):
 )
 
 
+@spmd.no_typecheck(out_types=spmd.PartitionSpec(("dp", "cp"), "tp", None))
+def varlen_attn_fn(*args, **kwargs):
```
Comment (tianyu-l) [inline torchtitan/models/common/attention.py]: nit: asymmetric naming, one with `_fn`, the other without

Can we `import varlen_attn as _varlen_attn` and name this one by `varlen_attn`?

File: torchtitan/distributed/spmd_types.py (hunk 5 above)
Comment (pianpwk) [reply torchtitan/distributed/spmd_types.py]: ah mistake here

File: torchtitan/models/common/attention.py [hunk 7]
```diff
@@ -175,47 +192,28 @@ def forward(
...
-                v_TNH.to(torch.bfloat16),
-                cu_seq_q,
-                cu_seq_k,
-                max_q,
-                max_k,
-                scale=scale,
-                window_size=self.window_size,
-                **varlen_kwargs,
-            )
+        run_varlen_attn = varlen_attn if out_transform is None else varlen_attn_with_lse
```
Comment (tianyu-l) [inline torchtitan/models/common/attention.py]: nit
```suggestion
        varlen_attn_fn = varlen_attn if out_transform is None else varlen_attn_with_lse
```

**Review by tianyu-l — APPROVED** [review]
